### DISCUSSION TITLE
How to approach and indicator chain that requires subtractive elements

### DISCUSSION THREAD

**Discussion by jimtollan** [DaveSkender/Stock.Indicators#977] [Ideas]
Hi Dave -me again!! seasons greetings. 

I've been looking at generating the MYC signal (see: https://www.singlebroker.com/blog/post/top-advanced-indicators-for-crypto-trading/) by combining/chaining the essential ingredients required to produce this single signal line. Below is the pseudo code required to create the indicator:

     static decimal[] CalculateMYC(decimal[] prices)
        {
            // Calculate the MACD for the prices
            var macd = CalculateMACD(prices, 12, 26, 9);

            // Calculate the 9-day EMA of the MACD
            var macdEma = CalculateEMA(macd, 9);

            // Calculate the 9-day SMA of the prices
            var ma = CalculateSMA(prices, 9);

            // Calculate the MYC Trading Indicator [psuedo list maths]
            var myc = macd - macdEma + ma;

            return myc;
        }

The main issue would be that if I chain together the current indicators, it would create an additive expression and would not take into account the macd - macdEma calculation. I could of course leave this as a method and be done with it, but just wondered if there was a way to go about this via chaining.

Hope the above makes sense.
jim

**Comment by jimtollan** [comment]
btw - seems that you may just be becoming an industry standard. saw this comparison page showing how this particular author's indicators stacked up against the rest: https://github.com/mihakralj/QuanTAlib ;)

**Reply by DaveSkender** [reply]
hehe, @mihakralj and I have been chatting for a while.  He’s also been contributing to this project in several ways to make it better, so we’re helping each other out.  After I get more of the roadmap done on this one, I’m hoping to collaborate with him more on new projects.

**Reply by mihakralj** [reply]
Dave's TA library is rock-solid and indeed an industry standard, along side with TA-LIB and Pandas-TA.

The requirement I had - and is not met in any of other libraries - was to allow incremental calculation for streaming (real-time) data feeds, either adjusting the last bar or forming a new bar. That's why QuanTAlib indicators are all based on an .Add() method, forcing every indicator to know how to calculate a single new value - instead of re-calculating the whole list just because there was one new data point added.

Plus using NaNs instead of nulls, CPU-native Double types and other tricks. When you have your own playground, you play by own rules. 😊 

As Dave said, long-term future is to somehow merge both efforts, including my current work on Portfolio class - a precursor for algo rules optimization and algo-trading.

**Reply by DaveSkender** [reply]
Yah, I’m workin’ up a streaming data preview now too and have an open draft pull request, but have a few scaling and chaining riddles to solve before settling on a design that’ll work well for more complex situations.  It’s a mind bender.

**Reply by mihakralj** [reply]
I solved chaining by using events - every `TSeries` is a publisher of its own events, and can also be a subscriber to all upstream (depending) `TSeries` events. Not super-elegant, but the pub-sub model works very well for streaming scenarios.

**Comment by DaveSkender** [comment]
Regarding your original idea here, I think there are some basic math utilities we can add.  The old `TA-LIB` had add, subtract, etc.  Would be fairly simple to implement and would be a good starter contribution for someone.  Please add an `enhancement` issue if you want to track it.

**Reply by jimtollan** [reply]
Hi Dave  - for me, it's not a hugely important enhancement as i can of course build around it for now, but in the bigger picture, it would add an additional layer of flexibility.

cheers again.. jim

**Reply by DaveSkender** [reply]
I've added this as issue #979 for tracking

**Reply by mihakralj** [reply]
There are two approaches: either create a new set of indicators for basic math functions (ADD, SUB, MUL, DIV), or create public static implicit operators for `Quote` class and define +,-,*,/ 

The second option (with static implicit operators) is more intuitive, but requires static Quotes list. The first option is used in `TA-LIB`, `Tulip` and `QuanTAlib` as it allows dynamic application of +,-,*,/ on streaming data objects.

**Reply by jimtollan** [reply]
would an extension method/chain help?? i.e.

    var newQuotes = prices.GetMacd()
                       .WithOperator(Enum.Minus).GetEma()
                       .WithOperator(Enum.Plus).GetSma();

Mad, bad or just sad?? :D

**Reply by DaveSkender** [reply]
I was thinking the same, but would have to dig in to check viability

**Reply by mihakralj** [reply]
Fluent interface (that allows chaining) is just a decorator that can be easily added.

The core of the problem is the logic of merging two time series CORRECTLY: what if length of series is not the same? What if timestamps don't match? How to handle merging of time series and scalars?

I spent ungodly anount of time making DIV(series, series) and SUB(series, series) work - and it is still not handling edge cases right...

**Reply by DaveSkender** [reply]
Some of our indicators like [Beta](https://dotnet.stockindicators.dev/indicators/Beta/#content) compare two series.  I've taken a rather draconian approach to it in the past to keep it simple, so it fails on two conditions 1) lengths don't match and 2) dates don't match.

I also have a utility that forces two series to match by adding missing entries in each other's series, but that leads to a bunch of `null` gaps, which may be okay here.
